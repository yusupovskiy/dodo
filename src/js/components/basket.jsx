import React, {Component} from 'react';
import { Link } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { selectFood } from '../actions/index'

import ListBasketFoods from '../containers/listBasketFoods';
import BasketAdditionalFoods from './basketAdditionalFoods';
import ButtonOrderRegistration from './buttonOrderRegistration';

class Basket extends Component {
  render() {
    return (
      <div id="open-basket" className="open-basket">
        <div className="panel-basket">
          <Link to="/" className="back"></Link>
          <h2 className="title title-main">Корзина</h2>
          <div id="basket" className="basket cool-scrollbar">
            <div className="list">
              <ListBasketFoods />
              <BasketAdditionalFoods />

            </div>
          </div>

          <ButtonOrderRegistration />
        </div>
      </div>
    );
  }
}

function mapStateToProps(state) {
  return {
    foods: state.foods,
    basketFoods: state.basketFoods,
  };
}
function matchDispatchToProps(dispatch) {
  return bindActionCreators({ selectFood: selectFood }, dispatch)
}

export default connect(mapStateToProps, matchDispatchToProps)(Basket);