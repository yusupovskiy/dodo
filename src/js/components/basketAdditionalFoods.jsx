import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { selectFood } from '../actions/index'

export class BasketAdditionalFoods extends Component {
  itemAdditionalGoods(foods) {
    return foods.map(food => {
      return (
        <li key={ food.id } className="item-additional-produc">
          <div className="product-additional-produc">
            <img className="image-additional-produc" src={ food.image } alt=" " />
            <div className="name-additional-produc">
              { food.name }<br />
              { food.price } ₽<br />
              <div className="btn__button" onClick={ () => this.props.selectFood(food, 'increase') }>В корзину</div>
            </div>
          </div>
        </li>
      );
    });
  }

  additionalFoods(section) {
    const foodsFilt = this.props.foods.filter(f => {
      const indexBaskerFood = this.props.basketFoods.findIndex(bf => bf.id === f.id);

      if(indexBaskerFood < 0 && f.category === section.category)
        return f;
    });

    if(foodsFilt.length > 0) {
      return (
        <section className="section-basket" key={ section.id }>
          <h3 className="title">{ section.title }</h3>

          <div className="additional-product cool-scrollbar">
            <ul className="box-additional-produc">

              { this.itemAdditionalGoods(foodsFilt) }

            </ul>
          </div>
        </section>
      );
    }
  }

  render() {
    return (
      <div className="additional-foods">
        { 
          this.props.additionalSections.map(s => {
            return this.additionalFoods(s);
          })
        }
      </div>
    );
  }
}

function mapStateToProps(state) {
  return {
    foods: state.foods,
    basketFoods: state.basketFoods,
    additionalSections: [
      { id: 1, title: 'Рекомендуем попробовать', category: 'pizza' },
      { id: 2, title: 'Напитки', category: 'drink' },
      { id: 3, title: 'Соусы', category: 'snacks' },
      { id: 4, title: 'Десерты', category: 'desserts' },
    ],
  };
}
function matchDispatchToProps(dispatch) {
  return bindActionCreators({
    selectFood
  }, dispatch)
}

export default connect(mapStateToProps, matchDispatchToProps)(BasketAdditionalFoods);