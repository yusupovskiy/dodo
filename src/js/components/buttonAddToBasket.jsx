import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { selectFood } from '../actions/index'

export class ButtonAddToBasket extends Component {
  render() {
    const id = this.props.id,
          indexBasketFood = this.props.basketFoods.findIndex(bf => bf.id === id),
          basketFood = this.props.basketFoods[indexBasketFood];

    if(basketFood !== undefined && basketFood.count > 0) {
      return(
        <div className="panel__count">
          <div className="btn__decrease btn__count" onClick={ () => this.props.selectFood(basketFood, 'decrease') }>-</div>
          <div className="text__count">{ basketFood.count }</div>
          <div className="btn__increase btn__count" onClick={ () => this.props.selectFood(basketFood, 'increase') }>+</div>
        </div>
      ); 
    }
    else {
      const food = this.props.foods.find(f => f.id === id);

      return(
        <div className="btn__button" onClick={ () => this.props.selectFood(food, 'increase') }>В корзину</div>
      ); 
    }
  }
}

function mapStateToProps(state) {
  return {
    foods: state.foods,
    basketFoods: state.basketFoods,
  };
}

function matchDispatchToProps(dispatch) {
  return bindActionCreators({ 
    selectFood
  }, dispatch)
}

export default connect(mapStateToProps, matchDispatchToProps)(ButtonAddToBasket);