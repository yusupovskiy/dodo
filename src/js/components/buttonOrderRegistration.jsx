import React, {Component} from 'react';
import { Link } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { selectFood } from '../actions/index'

export class ButtonOrderRegistration extends Component {
  orderRegistration() {
    this.props.selectFood({}, 'delete_all');
    alert("Спасибо за заказ, ожидайте!");
  }

  render() {
    if(this.props.basketFoods.length > 0) {
      return (
        <Link onClick={ () => { this.orderRegistration() } } to="/" className="btn">Оформить заказ</Link>
      );
    } else return (<div></div>);
  }
}

function mapStateToProps(state) {
  return {
    basketFoods: state.basketFoods,
  };
}
function matchDispatchToProps(dispatch) {
  return bindActionCreators({ 
    selectFood 
  }, dispatch)
}

export default connect(mapStateToProps, matchDispatchToProps)(ButtonOrderRegistration);