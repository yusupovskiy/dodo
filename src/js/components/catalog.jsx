import React from 'react';
import FoodsList from '../containers/listFoods';
import CatalogHeader from './catalogHeader';

const Catalog = () => (
  <div className="container">
    <CatalogHeader />
    <FoodsList />
  </div>
);

export default Catalog;