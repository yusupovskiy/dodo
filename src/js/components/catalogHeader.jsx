import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export class CatalogHeader extends Component {
  shortInfo() {
    const totalCountAndPrice = this.props.basketFoods.reduce((sum, food) => {
      const index = this.props.foods.findIndex(f => f.id === food.id),
            foodPrice = this.props.foods[index].price;

      if(sum <= 0) {
        sum = {
          count: food.count,
          price: (foodPrice * food.count)
        }
      }
      else {
        sum = {
          count: sum.count + food.count,
          price: sum.price + (foodPrice * food.count)
        }
      }

      return sum;
    }, 0);

    if(totalCountAndPrice !== 0) {
      return (
        <div className="short-info" onClick={ this.showBasket }>
          <div className="short-info_total-price">{ totalCountAndPrice.price } ₽</div>
          <div className="short-info__count">{ totalCountAndPrice.count } товаров</div>
        </div>
      );
    }
  }

  render() {
    return (
      <header>
        <a className="logo" href="/">
          <img src="https://dodopizza-a.akamaihd.net/site-static/dist/custom/8dbd4d5523ca05b07470.svg" />
        </a>

        <div className="menu">
          <a href="#pizza" className="item-menu">Пицца</a>
          <a href="#snacks" className="item-menu">Закуски</a>
          <a href="#desserts" className="item-menu">Десерты</a>
          <a href="#drink" className="item-menu">Напитки</a>
        </div>

        <Link className="btn__button" to="/basket" onClick={ this.showBasket }>Корзина</Link>

        { this.shortInfo() }
      </header>
    );
  }
}

function mapStateToProps(state) {
  return {
    foods: state.foods,
    basketFoods: state.basketFoods,
  };
}

export default connect(mapStateToProps)(CatalogHeader);
