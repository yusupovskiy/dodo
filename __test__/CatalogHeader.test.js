import '../setupTests'
import React from 'react'
import { CatalogHeader } from '../src/js/components/catalogHeader';

describe('Testion Component CatalogHeader', () => {
  const props = {
          foods: [{
              id: 1,
              name: "Жар-Баран",
              description: "Пищевая ценность продукта на 100 г",
              category: "pizza",
              ingredients: [{ name: "Томатный соус", main_ingredient: true }],
              price: 845, 
              image: "https://dodopizza-a.akamaihd.net/static/Img/Products/Pizza/ru-RU/49c9bd34-f725-400d-bca5-81e6c8bacc76.jpg",
            },
            {
              id: 2,
              name: "Сырный",
              description: "Пищевая ценность продукта на 100 г",
              category: "pizza",
              ingredients: [{ name: "Томатный соус", main_ingredient: true }],
              price: 845, 
              image: "https://dodopizza-a.akamaihd.net/static/Img/Products/Pizza/ru-RU/49c9bd34-f725-400d-bca5-81e6c8bacc76.jpg",
            }
          ],
          basketFoods: [{ id: 1, count: 3 }, { id: 2, count: 1 }],
        },
        wrapper = shallow(<CatalogHeader { ...props } />),
        totalCountAndPrice = props.basketFoods.reduce((sum, food) => {
          const index = props.foods.findIndex(f => f.id === food.id),
                foodPrice = props.foods[index].price;
    
          if(sum <= 0) {
            sum = {
              count: food.count,
              price: (foodPrice * food.count)
            }
          }
          else {
            sum = {
              count: sum.count + food.count,
              price: sum.price + (foodPrice * food.count)
            }
          }
    
          return sum;
        }, 0);
  
  it('SNAPSHOT Component header catalog', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Check whether the PRICE for all ordered food is summed up', () => {
    expect(wrapper
            .find('div.short-info_total-price')
            .text()
    ).toEqual(totalCountAndPrice.price + ' ₽');
  });

  it('Check whether the COUNT for all ordered food is summed up', () => {
    expect(wrapper
            .find('div.short-info__count')
            .text()
    ).toEqual(totalCountAndPrice.count + ' товаров');
  });
});