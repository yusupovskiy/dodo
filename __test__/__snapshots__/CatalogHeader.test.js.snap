// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testion Component CatalogHeader SNAPSHOT Component header catalog 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CatalogHeader
    basketFoods={
      Array [
        Object {
          "count": 3,
          "id": 1,
        },
        Object {
          "count": 1,
          "id": 2,
        },
      ]
    }
    foods={
      Array [
        Object {
          "category": "pizza",
          "description": "Пищевая ценность продукта на 100 г",
          "id": 1,
          "image": "https://dodopizza-a.akamaihd.net/static/Img/Products/Pizza/ru-RU/49c9bd34-f725-400d-bca5-81e6c8bacc76.jpg",
          "ingredients": Array [
            Object {
              "main_ingredient": true,
              "name": "Томатный соус",
            },
          ],
          "name": "Жар-Баран",
          "price": 845,
        },
        Object {
          "category": "pizza",
          "description": "Пищевая ценность продукта на 100 г",
          "id": 2,
          "image": "https://dodopizza-a.akamaihd.net/static/Img/Products/Pizza/ru-RU/49c9bd34-f725-400d-bca5-81e6c8bacc76.jpg",
          "ingredients": Array [
            Object {
              "main_ingredient": true,
              "name": "Томатный соус",
            },
          ],
          "name": "Сырный",
          "price": 845,
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <a
          className="logo"
          href="/"
        >
          <img
            src="https://dodopizza-a.akamaihd.net/site-static/dist/custom/8dbd4d5523ca05b07470.svg"
          />
        </a>,
        <div
          className="menu"
        >
          <a
            className="item-menu"
            href="#pizza"
          >
            Пицца
          </a>
          <a
            className="item-menu"
            href="#snacks"
          >
            Закуски
          </a>
          <a
            className="item-menu"
            href="#desserts"
          >
            Десерты
          </a>
          <a
            className="item-menu"
            href="#drink"
          >
            Напитки
          </a>
        </div>,
        <Link
          className="btn__button"
          replace={false}
          to="/basket"
        >
          Корзина
        </Link>,
        <div
          className="short-info"
        >
          <div
            className="short-info_total-price"
          >
            3380
             ₽
          </div>
          <div
            className="short-info__count"
          >
            4
             товаров
          </div>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <img
            src="https://dodopizza-a.akamaihd.net/site-static/dist/custom/8dbd4d5523ca05b07470.svg"
          />,
          "className": "logo",
          "href": "/",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "src": "https://dodopizza-a.akamaihd.net/site-static/dist/custom/8dbd4d5523ca05b07470.svg",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        "type": "a",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <a
              className="item-menu"
              href="#pizza"
            >
              Пицца
            </a>,
            <a
              className="item-menu"
              href="#snacks"
            >
              Закуски
            </a>,
            <a
              className="item-menu"
              href="#desserts"
            >
              Десерты
            </a>,
            <a
              className="item-menu"
              href="#drink"
            >
              Напитки
            </a>,
          ],
          "className": "menu",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Пицца",
              "className": "item-menu",
              "href": "#pizza",
            },
            "ref": null,
            "rendered": "Пицца",
            "type": "a",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Закуски",
              "className": "item-menu",
              "href": "#snacks",
            },
            "ref": null,
            "rendered": "Закуски",
            "type": "a",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Десерты",
              "className": "item-menu",
              "href": "#desserts",
            },
            "ref": null,
            "rendered": "Десерты",
            "type": "a",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Напитки",
              "className": "item-menu",
              "href": "#drink",
            },
            "ref": null,
            "rendered": "Напитки",
            "type": "a",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Корзина",
          "className": "btn__button",
          "onClick": undefined,
          "replace": false,
          "to": "/basket",
        },
        "ref": null,
        "rendered": "Корзина",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="short-info_total-price"
            >
              3380
               ₽
            </div>,
            <div
              className="short-info__count"
            >
              4
               товаров
            </div>,
          ],
          "className": "short-info",
          "onClick": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                3380,
                " ₽",
              ],
              "className": "short-info_total-price",
            },
            "ref": null,
            "rendered": Array [
              3380,
              " ₽",
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                4,
                " товаров",
              ],
              "className": "short-info__count",
            },
            "ref": null,
            "rendered": Array [
              4,
              " товаров",
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "header",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <a
            className="logo"
            href="/"
          >
            <img
              src="https://dodopizza-a.akamaihd.net/site-static/dist/custom/8dbd4d5523ca05b07470.svg"
            />
          </a>,
          <div
            className="menu"
          >
            <a
              className="item-menu"
              href="#pizza"
            >
              Пицца
            </a>
            <a
              className="item-menu"
              href="#snacks"
            >
              Закуски
            </a>
            <a
              className="item-menu"
              href="#desserts"
            >
              Десерты
            </a>
            <a
              className="item-menu"
              href="#drink"
            >
              Напитки
            </a>
          </div>,
          <Link
            className="btn__button"
            replace={false}
            to="/basket"
          >
            Корзина
          </Link>,
          <div
            className="short-info"
          >
            <div
              className="short-info_total-price"
            >
              3380
               ₽
            </div>
            <div
              className="short-info__count"
            >
              4
               товаров
            </div>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <img
              src="https://dodopizza-a.akamaihd.net/site-static/dist/custom/8dbd4d5523ca05b07470.svg"
            />,
            "className": "logo",
            "href": "/",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "src": "https://dodopizza-a.akamaihd.net/site-static/dist/custom/8dbd4d5523ca05b07470.svg",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "a",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <a
                className="item-menu"
                href="#pizza"
              >
                Пицца
              </a>,
              <a
                className="item-menu"
                href="#snacks"
              >
                Закуски
              </a>,
              <a
                className="item-menu"
                href="#desserts"
              >
                Десерты
              </a>,
              <a
                className="item-menu"
                href="#drink"
              >
                Напитки
              </a>,
            ],
            "className": "menu",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Пицца",
                "className": "item-menu",
                "href": "#pizza",
              },
              "ref": null,
              "rendered": "Пицца",
              "type": "a",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Закуски",
                "className": "item-menu",
                "href": "#snacks",
              },
              "ref": null,
              "rendered": "Закуски",
              "type": "a",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Десерты",
                "className": "item-menu",
                "href": "#desserts",
              },
              "ref": null,
              "rendered": "Десерты",
              "type": "a",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Напитки",
                "className": "item-menu",
                "href": "#drink",
              },
              "ref": null,
              "rendered": "Напитки",
              "type": "a",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Корзина",
            "className": "btn__button",
            "onClick": undefined,
            "replace": false,
            "to": "/basket",
          },
          "ref": null,
          "rendered": "Корзина",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="short-info_total-price"
              >
                3380
                 ₽
              </div>,
              <div
                className="short-info__count"
              >
                4
                 товаров
              </div>,
            ],
            "className": "short-info",
            "onClick": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  3380,
                  " ₽",
                ],
                "className": "short-info_total-price",
              },
              "ref": null,
              "rendered": Array [
                3380,
                " ₽",
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  4,
                  " товаров",
                ],
                "className": "short-info__count",
              },
              "ref": null,
              "rendered": Array [
                4,
                " товаров",
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "header",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
